### PR TITLE
Java: implement RANDOMKEY

### PR DESCRIPTION
**Pull request by cyip10**
*Issue #, if available:*
https://redis.io/docs/latest/commands/randomkey/

*Description of changes:*

Blocked on https://github.com/amazon-contributing/redis-rs/pull/153

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: java/client/src/main/java/glide/api/commands/GenericClusterCommands.java [hunk 1]
```diff
@@ -112,4 +112,38 @@ public interface GenericClusterCommands {
...
+     *
+     * @see <a href="https://redis.io/docs/latest/commands/randomkey/">redis.io</a> for details.
+     * @param route A single-node routing configuration for the transaction. The client will route the
+     *     transaction to the node defined by <code>route</code>.
+     * @return A random <code>key</code> from the database.
+     * @example
+     *     <pre>{@code
+     * String value = client.set("key", "value").get();
+     * String value_1 = client.set("key1", "value_1").get();
+     * String key = client.randomKey().get();
```
Comment (GumpacG) [inline java/client/src/main/java/glide/api/commands/GenericClusterCommands.java:127]: Can you add a route argument here?

File: java/integTest/src/test/java/glide/cluster/CommandTests.java [hunk 2]
```diff
@@ -817,4 +817,23 @@ public void functionLoad(boolean withRoute) {
...
+        assertEquals(OK, clusterClient.set(key1, "a").get());
+        assertEquals(OK, clusterClient.set(key2, "b").get());
+
+        String randomKey = clusterClient.randomKey().get();
+        assertEquals(1L, clusterClient.exists(new String[] {randomKey}).get());
+
+        // no keys in database
+        assertEquals(OK, clusterClient.flushall().get());
+        ExecutionException executionException =
+                assertThrows(ExecutionException.class, () -> clusterClient.randomKey().get());
```
Comment (GumpacG) [inline java/integTest/src/test/java/glide/cluster/CommandTests.java:836]: According to docs, shouldn't the return value be null?

Comment (cyip10) [reply java/integTest/src/test/java/glide/cluster/CommandTests.java:836]: it should be, however, when I do an assertNull, the value the program gives is "no value found- ResponseError"

File: java/integTest/src/test/java/glide/cluster/CommandTests.java [hunk 3]
```diff
@@ -817,4 +817,23 @@ public void functionLoad(boolean withRoute) {
...
+
+    @Test
+    @SneakyThrows
+    public void randomKey() {
+        String key1 = "{key}" + UUID.randomUUID();
+        String key2 = "{key}" + UUID.randomUUID();
+
+        assertEquals(OK, clusterClient.set(key1, "a").get());
+        assertEquals(OK, clusterClient.set(key2, "b").get());
+
```
Comment (Yury-Fridlyand) [inline java/integTest/src/test/java/glide/cluster/CommandTests.java:829]: Add tests for cluster with single node route and multi node route.
Add tests for standalone including switching between DBs - empty and nonempty ones.

File: java/integTest/src/test/java/glide/cluster/CommandTests.java (hunk 2 above)
Comment (Yury-Fridlyand) [reply java/integTest/src/test/java/glide/cluster/CommandTests.java:836]: Oh there is a bug in https://github.com/Bit-Quill/glide-for-redis/blob/f2c3d49e5961c890b13aa43e4b520bfb903173a6/glide-core/src/client/standalone_client.rs#L257-L270
You need to fix that in scope of this PR

File: java/client/src/main/java/glide/api/RedisClusterClient.java [hunk 4]
```diff
@@ -446,4 +447,16 @@ public CompletableFuture<String> functionLoadReplace(
                 route,
                 this::handleStringResponse);
     }
+
+    @Override
+    public CompletableFuture<String> randomKey(@NonNull SingleNodeRoute route) {
```
[the remark this answers is not in the document]
Comment (acarbonetto) [reply java/client/src/main/java/glide/api/RedisClusterClient.java:452]: You get the same result from any node.  So there's no reason to ever send this to multiple nodes.

File: java/integTest/src/test/java/glide/cluster/CommandTests.java (hunk 2 above)
Comment (acarbonetto) [reply java/integTest/src/test/java/glide/cluster/CommandTests.java:836]: I think we need to raise a redis-rs fix for randomKey that should return a null if there are no keys available:
https://github.com/amazon-contributing/redis-rs/blob/main/redis/src/cluster_routing.rs#L321-L322

Change to:
```
            // for randomkey, see: 
            // This isn't based on response_tips, but on the discussion here - https://github.com/redis/redis/issues/12410
            b"FUNCTION KILL" | b"SCRIPT KILL" | b"RANDOMKEY" => Some(OneSucceeded),
```

Comment (acarbonetto) [reply java/integTest/src/test/java/glide/cluster/CommandTests.java:836]: For now, please add a `// TODO` comment that says we need to update redis-rs to return a `OneSucceeded`

Comment (acarbonetto) [reply java/integTest/src/test/java/glide/cluster/CommandTests.java:840]: see: https://github.com/amazon-contributing/redis-rs/pull/149
Feel free to add this link as a TODO comment

File: java/client/src/main/java/glide/api/commands/GenericClusterCommands.java [hunk 5]
```diff
@@ -112,4 +112,37 @@ public interface GenericClusterCommands {
      * }</pre>
      */
     CompletableFuture<Object[]> exec(ClusterTransaction transaction, SingleNodeRoute route);
+
+    /**
+     * Return a random key.
+     *
+     * @see <a href="https://redis.io/docs/latest/commands/randomkey/">redis.io</a> for details.
+     * @param route
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/commands/GenericClusterCommands.java:120]: Add (copy) description

File: java/client/src/main/java/glide/api/commands/GenericClusterCommands.java [hunk 6]
```diff
@@ -112,4 +112,37 @@ public interface GenericClusterCommands {
      * }</pre>
      */
     CompletableFuture<Object[]> exec(ClusterTransaction transaction, SingleNodeRoute route);
+
+    /**
+     * Return a random key.
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/commands/GenericClusterCommands.java:117]:
```suggestion
     * Returns a random key.
```

File: java/client/src/main/java/glide/api/commands/GenericClusterCommands.java [hunk 7]
```diff
@@ -112,4 +112,37 @@ public interface GenericClusterCommands {
...
+     * String value_1 = client.set("key1", "value_1").get();
+     * String key = client.randomKey().get();
+     * System.out.println("The random key is: " + key);
+     * // The value of key is either "key" or "key1"
+     * }</pre>
+     */
+    CompletableFuture<String> randomKey(SingleNodeRoute route);
+
+    /**
+     * Return a random key.
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/commands/GenericClusterCommands.java:134]:
```suggestion
     * Returns a random key.
```

File: java/client/src/main/java/glide/api/commands/GenericCommands.java [hunk 8]
```diff
@@ -54,4 +54,20 @@ public interface GenericCommands {
      * }</pre>
      */
     CompletableFuture<Object[]> exec(Transaction transaction);
+
+    /**
+     * Return a random key.
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/commands/GenericCommands.java:59]:
```suggestion
     * Returns a random key from currently selected database.
```

File: java/integTest/src/test/java/glide/cluster/CommandTests.java [hunk 9]
```diff
@@ -817,4 +817,27 @@ public void functionLoad(boolean withRoute) {
...
+        String randomKey = clusterClient.randomKey().get();
+        assertEquals(1L, clusterClient.exists(new String[] {randomKey}).get());
+
+        // no keys in database
+        assertEquals(OK, clusterClient.flushall().get());
+        // TODO need to update redis-rs to return a OneSucceeded
+        // for randomkey, see:
+        // This isn't based on response_tips, but on the discussion here - https://github.com/redis/redis/issues/12410
+        // b"FUNCTION KILL" | b"SCRIPT KILL" | b"RANDOMKEY" => Some(OneSucceeded),
```
Comment (Yury-Fridlyand) [inline java/integTest/src/test/java/glide/cluster/CommandTests.java:838]: Put PR link in the TODO comment

File: java/integTest/src/test/java/glide/standalone/CommandTests.java [hunk 10]
```diff
@@ -364,4 +364,6 @@ public void functionLoad() {
         assertEquals(libName, regularClient.functionLoadReplace(newCode).get());
         // TODO test with FCALL
     }
+
+    // regualr client
```
Comment (Yury-Fridlyand) [inline java/integTest/src/test/java/glide/standalone/CommandTests.java:368]: TODO

File: java/client/src/main/java/glide/api/RedisClusterClient.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (GumpacG) [reply java/client/src/main/java/glide/api/RedisClusterClient.java:452]: So does that mean we can only use `RANDOM` from https://github.com/aws/glide-for-redis/blob/9caf33d11f660bf2b13f41975b2f44530a3bcf3c/java/client/src/main/java/glide/api/models/configuration/RequestRoutingConfiguration.java#L35?

[the remark this answers is not in the document]
Comment (acarbonetto) [reply java/client/src/main/java/glide/api/RedisClusterClient.java:452]: you can use `SlotIdRoute`, `SlotKeyRoute` and `ByAddressRoute`, which all implement `SingleNodeRoute`

Comment (acarbonetto) [inline java/client/src/main/java/glide/api/RedisClusterClient.java:452]: @cyip10 after further experimentation, we should change this to map to anything, since users should know what they're doing when they pick a route.  The ALL_PRIMARIES (default) is probably the best solution in 99.9% of the cases. 
```suggestion
    public CompletableFuture<String> randomKey(@NonNull Route route) {
```

File: java/client/src/main/java/glide/api/commands/GenericClusterCommands.java [hunk 11]
```diff
@@ -112,4 +112,38 @@ public interface GenericClusterCommands {
...
+     * String key = client.randomKey(RANDOM).get();
+     * System.out.println("The random key is: " + key);
+     * // The value of key is either "key" or "key1"
+     * }</pre>
+     */
+    CompletableFuture<String> randomKey(SingleNodeRoute route);
+
+    /**
+     * Returns a random key.
+     *
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/commands/GenericClusterCommands.java:136]: Need to add a note that specified that the command goes to all_primaries by default: 
```
The command will be routed to all primary nodes, and will return the first successful result. 
```

File: java/client/src/main/java/glide/api/commands/GenericClusterCommands.java [hunk 12]
```diff
@@ -112,4 +112,38 @@ public interface GenericClusterCommands {
      * }</pre>
      */
     CompletableFuture<Object[]> exec(ClusterTransaction transaction, SingleNodeRoute route);
+
+    /**
+     * Returns a random key.
+     *
+     * @see <a href="https://redis.io/docs/latest/commands/randomkey/">redis.io</a> for details.
+     * @param route Specifies the routing configuration for the command. The client will route the
+     *     command to the nodes defined by <code>route</code>.
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/commands/GenericClusterCommands.java:121]:
```suggestion
     *     command to the nodes defined by <code>route</code>, and will return the first successful result.
```

File: java/client/src/main/java/glide/api/commands/GenericCommands.java [hunk 13]
```diff
@@ -54,4 +54,20 @@ public interface GenericCommands {
      * }</pre>
      */
     CompletableFuture<Object[]> exec(Transaction transaction);
+
+    /**
+     * Returns a random key from currently selected database..
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/commands/GenericCommands.java:59]:
```suggestion
     * Returns a random key from currently selected database.
```

File: java/client/src/test/java/glide/api/RedisClientTest.java [hunk 14]
```diff
@@ -4050,6 +4051,28 @@ public void type_returns_success() {
         assertEquals(value, payload);
     }
 
+    @SneakyThrows
+    @Test
+    public void randomKey() {
+        // setup
+        String key1 = "key1";
+        String key2 = "key2";
+        String[] arguments = new String[] {};
```
Comment (acarbonetto) [inline java/client/src/test/java/glide/api/RedisClientTest.java:4060]: creating and saving an empty string doesn't really add any value.

File: java/client/src/test/java/glide/api/RedisClientTest.java [hunk 15]
```diff
@@ -4050,6 +4051,28 @@ public void type_returns_success() {
...
+        CompletableFuture<String> testResponse = new CompletableFuture<>();
+        testResponse.complete(key1);
+
+        // match on protobuf request
+        when(commandManager.<String>submitNewCommand(eq(RandomKey), eq(arguments), any()))
```
Comment (acarbonetto) [inline java/client/src/test/java/glide/api/RedisClientTest.java:4065]:
```suggestion
        when(commandManager.<String>submitNewCommand(eq(RandomKey), eq(new String[0]), any()))
```

File: java/client/src/test/java/glide/api/RedisClientTest.java [hunk 16]
```diff
@@ -4050,6 +4051,28 @@ public void type_returns_success() {
...
+                .thenReturn(testResponse);
+
+        // exercise
+        service.set(key1, "a");
```
Comment (acarbonetto) [inline java/client/src/test/java/glide/api/RedisClientTest.java:4069]: please remove.  this won't do anything.

File: java/client/src/test/java/glide/api/RedisClientTest.java [hunk 17]
```diff
@@ -4050,6 +4051,28 @@ public void type_returns_success() {
         assertEquals(value, payload);
     }
 
+    @SneakyThrows
+    @Test
+    public void randomKey() {
+        // setup
+        String key1 = "key1";
+        String key2 = "key2";
```
Comment (acarbonetto) [inline java/client/src/test/java/glide/api/RedisClientTest.java:4059]: we don't use key2 - please remove

File: java/client/src/main/java/glide/api/commands/GenericClusterCommands.java [hunk 18]
```diff
@@ -112,4 +112,40 @@ public interface GenericClusterCommands {
      * }</pre>
      */
     CompletableFuture<Object[]> exec(ClusterTransaction transaction, SingleNodeRoute route);
+
+    /**
+     * Returns a random key. The command will be routed to all primary nodes, and will return the
+     * first successful result.
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/commands/GenericClusterCommands.java:118]: Sorry, please include a break.  
```
     * Returns a random key. <br>
     * The command will be routed to all primary nodes, and will return the first successful result.
```

File: java/client/src/main/java/glide/api/models/BaseTransaction.java [hunk 19]
```diff
@@ -2833,6 +2834,17 @@ public T type(@NonNull String key) {
         return getThis();
     }
 
+    /**
+     * Return a random key from the currently selected database.
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/models/BaseTransaction.java:2838]:
```suggestion
     * Returns a random key from the currently selected database.
```

File: java/integTest/src/test/java/glide/TransactionTestUtilities.java [hunk 20]
```diff
@@ -184,6 +185,7 @@ private static Object[] stringCommands(BaseTransaction<?> transaction) {
 
         return new Object[] {
             OK, // set(stringKey1, value1)
+            stringKey1, // randomKey()
```
Comment (acarbonetto) [inline java/integTest/src/test/java/glide/TransactionTestUtilities.java:188]: this could be flaky. Since we're using this database for all tests, we could get in the situation where there's more than one key, and this returns something else.  In which case, the test would fail. 

Can you flush the database before the set, to guarantee that the database is empty...?  I think there might be a chance that another test adds another key while the transaction is processing, but I would think that would be very rare.

File: java/client/src/main/java/glide/api/models/BaseTransaction.java (hunk 19 above)
Comment (acarbonetto) [reply java/client/src/main/java/glide/api/models/BaseTransaction.java:2838]: - [ ] update doc

File: java/client/src/main/java/glide/api/commands/GenericClusterCommands.java (hunk 11 above)
Comment (acarbonetto) [reply java/client/src/main/java/glide/api/commands/GenericClusterCommands.java:138]: - [ ] update doc

File: java/client/src/main/java/glide/api/commands/GenericClusterCommands.java (hunk 18 above)
Comment (acarbonetto) [reply java/client/src/main/java/glide/api/commands/GenericClusterCommands.java:118]: - [ ] update doc

File: java/integTest/src/test/java/glide/TransactionTestUtilities.java (hunk 20 above)
Comment (acarbonetto) [reply java/integTest/src/test/java/glide/TransactionTestUtilities.java:188]: - [ ] harden test

**Review by acarbonetto — APPROVED** [review]
Approved pending the minor final comments.  
Please also rebase and try and run with the latest submodule commit. I think we want to make sure we don't check in redis-rs update and go backwards.

File: java/client/src/test/java/glide/api/RedisClientTest.java [hunk 21]
```diff
@@ -4050,6 +4051,23 @@ public void type_returns_success() {
...
+    @SneakyThrows
+    @Test
+    public void randomKey() {
+        // setup
+        String key1 = "key1";
+        CompletableFuture<String> testResponse = new CompletableFuture<>();
+        testResponse.complete(key1);
+
+        // match on protobuf request
+        when(commandManager.<String>submitNewCommand(eq(RandomKey), eq(new String[0]), any()));
```
Comment (acarbonetto) [inline java/client/src/test/java/glide/api/RedisClientTest.java:4063]:
```suggestion
        when(commandManager.<String>submitNewCommand(eq(RandomKey), eq(new String[0]), any())).thenReturn...;
```

File: java/client/src/test/java/glide/api/RedisClientTest.java [hunk 22]
```diff
@@ -4050,6 +4051,23 @@ public void type_returns_success() {
         assertEquals(value, payload);
     }
 
+    @SneakyThrows
+    @Test
+    public void randomKey() {
```
Comment (acarbonetto) [inline java/client/src/test/java/glide/api/RedisClientTest.java:4056]: you will also need to add tests for: 

- [ ] randomKey(Route route)
- [ ] randomKey()

in `RedisClusterClientTest`
